rust/lance/src/dataset/write/merge_insert.rs

```diff
@@ -0,0 +1,724 @@
+                        match fragment.extend_deletions(*bitmap).await {
+                            Ok(Some(new_fragment)) => {
+                                Ok(FragmentChange::Modified(new_fragment.metadata))
+                            }
+                            Ok(None) => Ok(FragmentChange::Removed(fragment_id as u64)),
+                            Err(e) => Err(e),
+                        }
+                    } else {
+                        Ok(FragmentChange::Unchanged)
+                    }
+                }
+            })
+            .buffer_unordered(num_cpus::get() * 4);
+
+        while let Some(res) = stream.next().await.transpose()? {
+            match res {
+                FragmentChange::Unchanged => {}
+                FragmentChange::Modified(fragment) => updated_fragments.push(fragment),
+                FragmentChange::Removed(fragment_id) => removed_fragments.push(fragment_id),
+            }
+        }
+
+        Ok((updated_fragments, removed_fragments))
+    }
+
+    // Commit the operation
+    async fn commit(
+        dataset: Arc<Dataset>,
+        removed_fragment_ids: Vec<u64>,
+        updated_fragments: Vec<Fragment>,
+        new_fragments: Vec<Fragment>,
+    ) -> Result<Arc<Dataset>> {
+        let operation = Operation::Update {
+            removed_fragment_ids,
+            updated_fragments,
+            new_fragments,
+        };
+        let transaction = Transaction::new(dataset.manifest.version, operation, None);
+
+        let manifest = commit_transaction(
+            dataset.as_ref(),
+            dataset.object_store(),
+            &transaction,
+            &Default::default(),
+            &Default::default(),
+        )
+        .await?;
+
+        let mut dataset = dataset.as_ref().clone();
+        dataset.manifest = Arc::new(manifest);
+
+        Ok(Arc::new(dataset))
+    }
+}
+
+// A sync-safe structure that is shared by all of the "process batch" tasks.
+//
+// Note: we are not currently using parallelism but this still needs to be sync because it is
+//       held across an await boundary (and we might user parallelism someday)
+#[derive(Debug, Clone)]
+struct Merger {
+    // As the merger runs it will update the list of deleted rows
+    deleted_rows: Arc<Mutex<RoaringTreemap>>,
+    // Physical delete expression, only set if params.delete_not_matched_by_source is DeleteIf
+    delete_expr: Option<Arc<dyn PhysicalExpr>>,
+    // The parameters controlling the merge
+    params: MergeInsertParams,
+}
+
+impl Merger {
+    // Creates a new merger with an empty set of deleted rows, compiles expressions, if present
+    fn try_new(params: MergeInsertParams, schema: Arc<Schema>) -> Result<Self> {
+        let delete_expr = if let WhenNotMatchedBySource::DeleteIf(expr) =
+            &params.delete_not_matched_by_source
+        {
+            let planner = Planner::new(schema.clone());
+            let physical_expr = planner.create_physical_expr(expr)?;
```

Comment:
Note: you do have to ensure that `planner.optimize_expr()` is run before `create_physical_expr`, otherwise the latter might error due to type mismatches.

Reply:
Thanks, I call `optimize_expr` now.